This depends on the hash changes of #25614; without it, `Interval(2, 3.0).intersection(Interval(2.0, 3))` fails while none of the other permutations of float amongst the args do.

When equal, the intersection routine does not run
```python
>>> Interval(2,3).intersection(Interval(2,3))
Interval(2, 3)
>>> Interval(2.,3.).intersection(Interval(2.,3.))
Interval(2.00000000000000, 3.00000000000000)
```
Only one pattern fails when mixing float/int args:
```python
>>> list(cartes([2,2.],[3,3.]))
[(2, 3), (2, 3.0), (2.0, 3), (2.0, 3.0)]
>>> for i in subsets(_,2):
...  try: Interval(*i[0]).intersection(Interval(*i[1]))
...  except: print('fail',i)
...
fail ((2, 3.0), (2.0, 3))
```
And this is the error
```python
>>> Interval(2.,3).intersection(Interval(2,3.))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "\sympy\sympy\sets\sets.py", line 156, in intersection
    return self.intersect(other)
  File "\sympy\sympy\sets\sets.py", line 150, in intersect
    return Intersection(self, other)
  File "\sympy\sympy\sets\sets.py", line 1502, in __new__
    args = list(ordered(set(_sympify(args))))
  File "\sympy\sympy\core\sorting.py", line 309, in ordered
    yield from value
  File "\sympy\sympy\core\sorting.py", line 309, in ordered
    yield from value
  File "\sympy\sympy\core\sorting.py", line 297, in ordered
    for k, value in sorted(d.items()):
TypeError: '<' not supported between instances of 'Interval' and 'Interval'
```
When I add in this diff, it works
```diff
diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index 452feeb49e..39692170d4 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -1101,37 +1101,22 @@ def __int__(self):
         return int(mlib.to_int(self._mpf_))  # uses round_fast = round_down
 
     def __eq__(self, other):
-        from sympy.logic.boolalg import Boolean
-        try:
-            other = _sympify(other)
-        except SympifyError:
-            return NotImplemented
-        if isinstance(other, Boolean):
-            return False
-        if other.is_NumberSymbol:
-            if other.is_irrational:
-                return False
-            return other.__eq__(self)
-        if other.is_Float:
-            # comparison is exact
-            # so Float(.1, 3) != Float(.1, 33)
-            return self._mpf_ == other._mpf_
-        if other.is_Rational:
-            return other.__eq__(self)
-        if other.is_Number:
-            # numbers should compare at the same precision;
-            # all _as_mpf_val routines should be sure to abide
-            # by the request to change the prec if necessary; if
-            # they don't, the equality test will fail since it compares
-            # the mpf tuples
-            ompf = other._as_mpf_val(self._prec)
-            return bool(mlib.mpf_eq(self._mpf_, ompf))
-        if not self:
-            return not other
-        return False    # Float != non-Number
+        if isinstance(other, float):
+            other = Float(other)
+        return Basic.__eq__(self, other)
 
     def __ne__(self, other):
-        return not self == other
+        eq = self.__eq__(other)
+        if eq is NotImplemented:
+            return eq
+        else:
+            return not eq
+
+    def __hash__(self):
+        float_val = float(self)
+        if not math.isinf(float_val):
+            return hash(float_val)
+        return Basic.__hash__(self)
 
     def _Frel(self, other, op):
         try:
@@ -1194,9 +1179,6 @@ def __le__(self, other):
             return Expr.__le__(self, other)
         return rv
 
-    def __hash__(self):
-        return super().__hash__()
-
     def epsilon_eq(self, other, epsilon="1e-15"):
         return abs(self - other) < Float(epsilon)
 
```
But that is the whole point of the other PR, so I can pull that out as a sequel instead of a prelude to #25614.

--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -2113,18 +2113,21 @@ def _aresame(a, b):
     Examples
     ========
 
-    In SymPy (as in Python) two numbers compare the same if they
-    have the same underlying base-2 representation even though
-    they may not be the same type:
+    In SymPy (unlike Python) two numbers do not compare the same if they are
+    not of the same type. In particular an Integer or Rational will always
+    compare unequal with a Float. Also Floats of differing precision will
+    compare unequal.
 
     >>> from sympy import S
     >>> 2.0 == S(2)
-    True
+    False
     >>> 0.5 == S.Half
-    True
+    False
 
-    This routine was written to provide a query for such cases that
-    would give false when the types do not match:
+    Before SymPy 1.7 the two comparisons above would have given True rather
+    than False. This routine was written to provide a query that could
+    distinguish those cases but is no longer needed after the changes to Float
+    comparison introduced in SymPy 1.7. Its usage is
 
     >>> from sympy.core.basic import _aresame
     >>> _aresame(S(2.0), S(2))

--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -14,8 +14,7 @@ from .expr import Expr, AtomicExpr
 from .evalf import pure_complex
 from .cache import cacheit, clear_cache
 from .decorators import _sympifyit
-from .intfunc import (num_digits, igcd, ilcm, mod_inverse,
-    integer_nthroot, integer_log)
+from .intfunc import num_digits, igcd, ilcm, mod_inverse, integer_nthroot
 from .logic import fuzzy_not
 from .kind import NumberKind
 from sympy.external.gmpy import SYMPY_INTS, gmpy, flint

--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -4166,7 +4166,7 @@ def equal_valued(x, y):
     However an individual Float compares equal to a Rational:
 
     >>> Rational(1, 2) == Float(0.5)
-    True
+    False
 
     In a future version of SymPy this might change so that Rational and Float
     compare unequal. This function provides the behavior currently expected of

--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -1044,7 +1044,7 @@ class Float(Number):
                          precision=self._prec)
         if isinstance(other, Float) and global_parameters.evaluate:
             r = self/other
-            if r == int(r):
+            if r == r.round(0):
                 return Float(0, precision=max(self._prec, other._prec))
         if isinstance(other, Number) and global_parameters.evaluate:
             rhs, prec = other._as_mpf_op(self._prec)
@@ -1101,37 +1101,22 @@ class Float(Number):
         return int(mlib.to_int(self._mpf_))  # uses round_fast = round_down
 
     def __eq__(self, other):
-        from sympy.logic.boolalg import Boolean
-        try:
-            other = _sympify(other)
-        except SympifyError:
-            return NotImplemented
-        if isinstance(other, Boolean):
-            return False
-        if other.is_NumberSymbol:
-            if other.is_irrational:
-                return False
-            return other.__eq__(self)
-        if other.is_Float:
-            # comparison is exact
-            # so Float(.1, 3) != Float(.1, 33)
-            return self._mpf_ == other._mpf_
-        if other.is_Rational:
-            return other.__eq__(self)
-        if other.is_Number:
-            # numbers should compare at the same precision;
-            # all _as_mpf_val routines should be sure to abide
-            # by the request to change the prec if necessary; if
-            # they don't, the equality test will fail since it compares
-            # the mpf tuples
-            ompf = other._as_mpf_val(self._prec)
-            return bool(mlib.mpf_eq(self._mpf_, ompf))
-        if not self:
-            return not other
-        return False    # Float != non-Number
+        if isinstance(other, float):
+            other = Float(other)
+        return Basic.__eq__(self, other)
 
     def __ne__(self, other):
-        return not self == other
+        eq = self.__eq__(other)
+        if eq is NotImplemented:
+            return eq
+        else:
+            return not eq
+
+    def __hash__(self):
+        float_val = float(self)
+        if not math.isinf(float_val):
+            return hash(float_val)
+        return Basic.__hash__(self)
 
     def _Frel(self, other, op):
         try:
@@ -1193,9 +1178,6 @@ class Float(Number):
         if rv is None:
             return Expr.__le__(self, other)
         return rv
-
-    def __hash__(self):
-        return super().__hash__()
 
     def epsilon_eq(self, other, epsilon="1e-15"):
         return abs(self - other) < Float(epsilon)
@@ -1620,30 +1602,7 @@ class Rational(Number):
             # so we can just check equivalence of args
             return self.p == other.p and self.q == other.q
         if other.is_Float:
-            # all Floats have a denominator that is a power of 2
-            # so if self doesn't, it can't be equal to other
-            if self.q & (self.q - 1):
-                return False
-            s, m, t = other._mpf_[:3]
-            if s:
-                m = -m
-            if not t:
-                # other is an odd integer
-                if not self.is_Integer or self.is_even:
-                    return False
-                return m == self.p
-
-            if t > 0:
-                # other is an even integer
-                if not self.is_Integer:
-                    return False
-                # does m*2**t == self.p
-                return self.p and not self.p % m and \
-                    integer_log(self.p//m, 2) == (t, True)
-            # does non-integer s*m/2**-t = p/q?
-            if self.is_Integer:
-                return False
-            return m == self.p and integer_log(self.q, 2) == (-t, True)
+            return False
         return False
 
     def __ne__(self, other):

--- a/sympy/core/tests/test_expr.py
+++ b/sympy/core/tests/test_expr.py
@@ -2015,6 +2015,11 @@ def test_round():
     assert a.round(26) == Float('3.000000000000000000000000000', '')
     assert a.round(27) == Float('2.999999999999999999999999999', '')
     assert a.round(30) == Float('2.999999999999999999999999999', '')
+    #assert a.round(10) == Float('3.0000000000', '')
+    #assert a.round(25) == Float('3.0000000000000000000000000', '')
+    #assert a.round(26) == Float('3.00000000000000000000000000', '')
+    #assert a.round(27) == Float('2.999999999999999999999999999', '')
+    #assert a.round(30) == Float('2.999999999999999999999999999', '')
 
     # XXX: Should round set the precision of the result?
     #      The previous version of the tests above is this but they only pass

--- a/sympy/core/tests/test_numbers.py
+++ b/sympy/core/tests/test_numbers.py
@@ -190,10 +190,6 @@ def test_divmod():
     assert divmod(S(-3), S(2)) == (-2, 1)
     assert divmod(S(-3), 2) == (-2, 1)
 
-    assert divmod(S(4), S(-3.1)) == Tuple(-2, -2.2)
-    # divmod(4, -2.1) gives an inaccurate result for the remainder
-    # assert divmod(S(4), S(-2.1)) == divmod(4, -2.1)
-    assert divmod(S(-8), S(-2.5) ) == Tuple(3, -0.5)
 
     assert divmod(oo, 1) == (S.NaN, S.NaN)
     assert divmod(S.NaN, 1) == (S.NaN, S.NaN)

--- a/sympy/core/tests/test_numbers.py
+++ b/sympy/core/tests/test_numbers.py
@@ -190,6 +190,11 @@ def test_divmod():
     assert divmod(S(-3), S(2)) == (-2, 1)
     assert divmod(S(-3), 2) == (-2, 1)
 
+    assert divmod(S(4), S(-3.1)) == Tuple(-2, -2.2)
+    # divmod(4, -2.1) gives an inaccurate result for the remainder
+    # assert divmod(S(4), S(-2.1)) == divmod(4, -2.1)
+    assert divmod(S(-8), S(-2.5) ) == Tuple(3, -0.5)
+
     assert divmod(oo, 1) == (S.NaN, S.NaN)
     assert divmod(S.NaN, 1) == (S.NaN, S.NaN)
     assert divmod(1, S.NaN) == (S.NaN, S.NaN)
@@ -455,7 +460,9 @@ def test_Float():
         return (-t < a - b < t)
 
     zeros = (0, S.Zero, 0., Float(0))
-    for i, j in permutations(zeros, 2):
+    for i, j in permutations(zeros[:-1], 2):
+        assert i == j
+    for i, j in permutations(zeros[-2:], 2):
         assert i == j
     for z in zeros:
         assert z in zeros
@@ -529,8 +536,10 @@ def test_Float():
 
     # inexact floats (repeating binary = denom not multiple of 2)
     # cannot have precision greater than 15
-    assert Float(.125, 22) == .125
-    assert Float(2.0, 22) == 2
+    assert Float(.125, 22)._prec == 76
+    assert Float(2.0, 22)._prec == 76
+    #assert Float(.125, 22) == .125
+    #assert Float(2.0, 22) == 2
     assert float(Float('.12500000000000001', '')) == .125
     raises(ValueError, lambda: Float(.12500000000000001, ''))
 
@@ -1385,7 +1394,9 @@ def test_abs1():
 
 
 def test_accept_int():
-    assert Float(4) == 4
+    assert not Float(4) == 4
+    assert Float(4) != 4
+    assert Float(4) == 4.0
 
 
 def test_dont_accept_str():
@@ -1851,8 +1862,8 @@ def test_bool_eq():
 
 
 def test_Float_eq():
-    # all .5 values are the same
-    assert Float(.5, 10) == Float(.5, 11) == Float(.5, 1)
+    # Floats with different precision should not compare equal
+    assert Float(.5, 10) != Float(.5, 11) != Float(.5, 1)
     # but floats that aren't exact in base-2 still
     # don't compare the same because they have different
     # underlying mpf values
@@ -1868,23 +1879,27 @@ def test_Float_eq():
     assert Rational(11, 10) != Float('1.1')
     # coverage
     assert not Float(3) == 2
+    assert not Float(3) == Float(2)
+    assert not Float(3) == 3
     assert not Float(2**2) == S.Half
-    assert Float(2**2) == 4
+    assert Float(2**2) == 4.0
     assert not Float(2**-2) == 1
-    assert Float(2**-1) == S.Half
+    assert Float(2**-1) == 0.5
     assert not Float(2*3) == 3
-    assert not Float(2*3) == S.Half
-    assert Float(2*3) == 6
+    assert not Float(2*3) == 0.5
+    assert Float(2*3) == 6.0
+    assert not Float(2*3) == 6
     assert not Float(2*3) == 8
-    assert Float(.75) == Rational(3, 4)
+    assert not Float(.75) == Rational(3, 4)
+    assert Float(.75) == 0.75
     assert Float(5/18) == 5/18
     # 4473
     assert Float(2.) != 3
-    assert Float((0,1,-3)) == S.One/8
+    assert not Float((0,1,-3)) == S.One/8
+    assert Float((0,1,-3)) == 1/8
     assert Float((0,1,-3)) != S.One/9
     # 16196
-    assert 2 == Float(2)  # as per Python
-    # but in a computation...
+    assert not 2 == Float(2)  # unlike Python
     assert t**2 != t**2.0
 
 

--- a/sympy/holonomic/holonomic.py
+++ b/sympy/holonomic/holonomic.py
@@ -1965,7 +1965,7 @@ class HolonomicFunction:
             sol = S.Zero
             for j, i in enumerate(nonzeroterms):
 
-                if i < 0 or int(i) != i:
+                if i < 0 or not (int(i) == i or i.round(0) == i):
                     continue
 
                 i = int(i)

--- a/sympy/matrices/common.py
+++ b/sympy/matrices/common.py
@@ -15,6 +15,7 @@ from sympy.core.basic import Atom
 from sympy.core.decorators import call_highest_priority
 from sympy.core.kind import Kind, NumberKind
 from sympy.core.logic import fuzzy_and, FuzzyBool
+from sympy.core.numbers import Integer
 from sympy.core.mod import Mod
 from sympy.core.singleton import S
 from sympy.core.symbol import Symbol
@@ -2848,6 +2849,8 @@ class MatrixArithmetic(MatrixRequired):
             return a._eval_pow_by_recursion(exp)
 
         elif method is None and exp.is_Number and exp % 1 == 0:
+            if exp.is_Float:
+                exp = Integer(exp)
             # Decide heuristically which method to apply
             if a.rows == 2 and exp > 100000:
                 return jordan_pow(exp)

--- a/sympy/matrices/expressions/blockmatrix.py
+++ b/sympy/matrices/expressions/blockmatrix.py
@@ -505,13 +505,13 @@ class BlockMatrix(MatrixExpr):
             [[A, B],
              [C, D]] = self.blocks.tolist()
             try:
-                A = A**0.5
+                A = A**S.Half
                 AI = A.I
             except NonInvertibleMatrixError:
                 raise NonInvertibleMatrixError('Block LU decomposition cannot be calculated when\
                     "A" is singular')
             Z = ZeroMatrix(*B.shape)
-            Q = self.schur()**0.5
+            Q = self.schur()**S.Half
             L = BlockMatrix([[A, Z], [C*AI, Q]])
             U = BlockMatrix([[A, AI*B],[Z.T, Q]])
             return L, U

--- a/sympy/ntheory/primetest.py
+++ b/sympy/ntheory/primetest.py
@@ -616,10 +616,10 @@ def isprime(n):
     >>> near_int == int(near_int)
     False
     >>> n = Float(near_int, 10)  # truncated by precision
-    >>> n == int(n)
+    >>> n == n.round(0)
     True
     >>> n = Float(near_int, 20)
-    >>> n == int(n)
+    >>> n == n.round(0)
     False
 
     See Also

--- a/sympy/physics/mechanics/tests/test_kane2.py
+++ b/sympy/physics/mechanics/tests/test_kane2.py
@@ -186,8 +186,10 @@ def test_aux_dep():
 
     assert Matrix(Fr_c).expand() == fr.expand()
     assert Matrix(Fr_star_c.subs(kdd)).expand() == frstar.expand()
-    assert (simplify(Matrix(Fr_star_steady).expand()) ==
-            simplify(frstar_steady.expand()))
+    # These Matrices have some Integer(0) and some Float(0). Running under
+    # SymEngine gives different types of zero.
+    assert (simplify(Matrix(Fr_star_steady).expand()).xreplace({0:0.0}) ==
+            simplify(frstar_steady.expand()).xreplace({0:0.0}))
 
     syms_in_forcing = find_dynamicsymbols(kane.forcing)
     for qdi in qd:

--- a/sympy/series/tests/test_limits.py
+++ b/sympy/series/tests/test_limits.py
@@ -257,7 +257,7 @@ def test_frac():
     raises(ValueError, lambda: limit(frac(x), x, 0, '+-'))
     assert limit(frac(-2*x + 1), x, 0, "+") == 1
     assert limit(frac(-2*x + 1), x, 0, "-") == 0
-    assert limit(frac(x + S.Half), x, 0, "+-") == 1/2
+    assert limit(frac(x + S.Half), x, 0, "+-") == S(1)/2
     assert limit(frac(1/x), x, 0) == AccumBounds(0, 1)
 
 
@@ -421,10 +421,10 @@ def test_issue_4547():
 
 def test_issue_5164():
     assert limit(x**0.5, x, oo) == oo**0.5 is oo
-    assert limit(x**0.5, x, 16) == S(16)**0.5
+    assert limit(x**0.5, x, 16) == 4 # Should this be a float?
     assert limit(x**0.5, x, 0) == 0
     assert limit(x**(-0.5), x, oo) == 0
-    assert limit(x**(-0.5), x, 4) == S(4)**(-0.5)
+    assert limit(x**(-0.5), x, 4) == S.Half # Should this be a float?
 
 
 def test_issue_5383():

--- a/sympy/series/tests/test_series.py
+++ b/sympy/series/tests/test_series.py
@@ -233,10 +233,11 @@ def test_issue_12791():
     expr = (-beta**2*varphi*sin(theta) + beta**2*cos(theta) + \
         beta*varphi*sin(theta) - beta*cos(theta) - beta + 1)/(beta*cos(theta) - 1)**2
 
-    sol = 0.5/(0.5*cos(theta) - 1.0)**2 - 0.25*cos(theta)/(0.5*cos(theta)\
-        - 1.0)**2 + (beta - 0.5)*(-0.25*varphi*sin(2*theta) - 1.5*cos(theta)\
-        + 0.25*cos(2*theta) + 1.25)/(0.5*cos(theta) - 1.0)**3\
-        + 0.25*varphi*sin(theta)/(0.5*cos(theta) - 1.0)**2 + O((beta - S.Half)**2, (beta, S.Half))
+    sol = (0.5/(0.5*cos(theta) - 1.0)**2 - 0.25*cos(theta)/(0.5*cos(theta) - 1.0)**2
+        + (beta - 0.5)*(-0.25*varphi*sin(2*theta) - 1.5*cos(theta)
+        + 0.25*cos(2*theta) + 1.25)/((0.5*cos(theta) - 1.0)**2*(0.5*cos(theta) - 1.0))
+        + 0.25*varphi*sin(theta)/(0.5*cos(theta) - 1.0)**2
+        + O((beta - S.Half)**2, (beta, S.Half)))
 
     assert expr.series(beta, 0.5, 2).trigsimp() == sol
 

--- a/sympy/utilities/misc.py
+++ b/sympy/utilities/misc.py
@@ -560,6 +560,6 @@ def as_int(n, strict=True):
             result = int(n)
         except TypeError:
             raise ValueError('%s is not an integer' % (n,))
-        if n != result:
+        if not (n == result or n == n.round(0)):
             raise ValueError('%s is not an integer' % (n,))
         return result

